File: agent/src/bundles/ignore_rules.ts
// Path: agent/src/bundles/ignore_rules.ts
// Description: Centralized ignore patterns for bundle building

/**
 * Directories that are always excluded from bundles
 */
export const ALWAYS_IGNORE_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  "target",
  ".next",
  ".cache",
  "logs",
  ".turbo",
  ".vercel",
  "__pycache__",
  ".mypy_cache",
  ".pytest_cache",
  "coverage",
]);

/**
 * File patterns that are always excluded from bundles
 */
export const ALWAYS_IGNORE_FILES = new Set([
  ".DS_Store",
  "Thumbs.db",
  ".env",
  ".env.local",
]);

/**
 * Check if an entry should be ignored during bundle building
 */
export function shouldIgnoreEntry(name: string, isDirectory: boolean): boolean {
  // Skip hidden files/dirs (starting with .)
  if (name.startsWith(".") && name !== ".") {
    // Allow certain dotfiles if needed in future
    return true;
  }

  if (isDirectory) {
    return ALWAYS_IGNORE_DIRS.has(name);
  }

  return ALWAYS_IGNORE_FILES.has(name);
}

File: crates/im_bundle/src/ignore_rules.rs
// Path: crates/im_bundle/src/ignore_rules.rs
// Description: Always-ignored file and directory names for bundle scanning

const IGNORED_DIRS: &[&str] = &[
    ".cache",
    ".git",
    ".mypy_cache",
    ".next",
    ".nyc_output",
    ".nuxt",
    ".parcel-cache",
    ".pytest_cache",
    ".ruff_cache",
    ".svelte-kit",
    ".tox",
    ".turbo",
    ".venv",
    ".gradle",
    ".hypothesis",
    ".nox",
    "__pycache__",
    "build",
    "coverage",
    "dist",
    "env",
    "logs",
    "node_modules",
    "out",
    "target",
    "venv",
];

const IGNORED_DIR_SUFFIXES: &[&str] = &[".egg-info"];

const IGNORED_FILES: &[&str] = &[
    ".DS_Store",
    ".coverage",
    ".env",
    ".env.local",
    ".eslintcache",
    "Thumbs.db",
    "thumbs.db",
];

const IGNORED_FILE_SUFFIXES: &[&str] = &[
    ".bak",
    ".log",
    ".old",
    ".orig",
    ".pyc",
    ".pyd",
    ".pyo",
    ".swo",
    ".swp",
    ".tmp",
    "~",
];

pub fn is_ignored_dir(name: &str) -> bool {
    if IGNORED_DIRS.contains(&name) {
        return true;
    }
    IGNORED_DIR_SUFFIXES
        .iter()
        .any(|suffix| name.ends_with(suffix))
}

pub fn is_ignored_file(name: &str) -> bool {
    if IGNORED_FILES.contains(&name) {
        return true;
    }
    IGNORED_FILE_SUFFIXES
        .iter()
        .any(|suffix| name.ends_with(suffix))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ignore_rules_match() {
        assert!(is_ignored_dir("node_modules"));
        assert!(is_ignored_dir(".git"));
        assert!(is_ignored_dir("package.egg-info"));
        assert!(is_ignored_file(".env"));
        assert!(is_ignored_file("debug.log"));
        assert!(is_ignored_file("module.pyc"));
        assert!(!is_ignored_file("README.md"));
    }
}
